src/szk.rs

```diff
@@ -0,0 +1,444 @@
+pub struct SzkQueryShares<const SEED_SIZE: usize, F> {
+    joint_rand_seed_and_part: Option<[Seed<SEED_SIZE>; 2]>,
+    verifier: SzkVerifier<F>,
+}
+
+impl<const SEED_SIZE: usize, F> SzkQueryShares<SEED_SIZE, F> {
+    pub fn new(
+        joint_rand_seed_and_part: Option<[Seed<SEED_SIZE>; 2]>,
+        verifier: SzkVerifier<F>,
+    ) -> Self {
+        Self {
+            joint_rand_seed_and_part,
+            verifier,
+        }
+    }
+}
+
+/// Verifier type for Szk Proof
+pub type SzkVerifier<F> = Vec<F>;
+
+/// Main struct encapsulating the shared zero-knowledge functionality.
+/// The type T is the underlying FLP proof system.
+/// P is the XOF used to derive all random coins (it should be indifferentiable
+/// from a random oracle for security.)
+pub struct Szk<T, P, const SEED_SIZE: usize>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    typ: T,
+    algorithm_id: u32,
+    phantom: PhantomData<P>,
+}
+
+impl<T: Type> Szk<T, XofTurboShake128, 16> {
+    /// Create an instance of [`Szk`] using [`XofTurboShake128`].
+    pub fn new_turboshake128(typ: T, algorithm_id: u32) -> Self {
+        Szk::new(typ, algorithm_id)
+    }
+}
+
+impl<T, P, const SEED_SIZE: usize> Szk<T, P, SEED_SIZE>
+where
+    T: Type,
+    P: Xof<SEED_SIZE>,
+{
+    /// Construct an instance of this sharedZK proof system with the underlying FLP.
+    pub fn new(typ: T, algorithm_id: u32) -> Self {
+        Self {
+            typ,
+            algorithm_id,
+            phantom: PhantomData,
+        }
+    }
+
+    fn adjust_domsep_tag(&self, usage: u16) -> [u8; 8] {
+        let mut adjusted_dst = [0u8; 8];
+        adjusted_dst[0] = MASTIC_VERSION;
+        adjusted_dst[1] = 0; // algorithm class
+        adjusted_dst[2..6].copy_from_slice(&(self.algorithm_id).to_be_bytes());
+        adjusted_dst[6..8].copy_from_slice(&usage.to_be_bytes());
+        adjusted_dst
+    }
+
+    /// Derive a vector of random field elements for consumption by the FLP prover.
+    fn derive_prove_rands(&self, prove_rand_seed: &Seed<SEED_SIZE>) -> Vec<T::Field> {
+        P::seed_stream(
+            prove_rand_seed,
+            &self.adjust_domsep_tag(DST_PROVE_RANDOMNESS),
+            &[],
+        )
+        .into_field_vec(self.typ.prove_rand_len())
+    }
+
+    fn derive_joint_rand_part(
+        &self,
+        aggregator_blind: &Seed<SEED_SIZE>,
+        measurement_share: &[T::Field],
+        nonce: &[u8; 16],
+    ) -> Result<Seed<SEED_SIZE>, SzkError> {
+        let mut xof = P::init(
+            aggregator_blind.as_ref(),
+            &self.adjust_domsep_tag(DST_JOINT_RAND_PART),
+        );
+        // Encode measurement_share (currently an array of field elements) into bytes
+        // and include it in the XOF input.
+        let mut bytes = vec![0u8; T::Field::ENCODED_SIZE];
+        for field_elem in measurement_share {
+            field_elem.encode(&mut bytes)?;
+            xof.update(&bytes)
+        }
+        xof.update(nonce);
```

Comment:
Proving robustness might require requires extracting the nonce from the attacker's RO query. In that spirit, it's a good idea to set up inputs to the XOF so that we can decode the individual components. I think we're fine in this case since we always know the length of `measurement_share`. Still, I might hash the variable length thing last.